internal/build/buildkit_printer.go

```diff
@@ -10,84 +10,112 @@ import (
 
 type buildkitPrinter struct {
 	output io.Writer
-	vMap   map[digest.Digest]*vertexAndLogs
+	vData  map[digest.Digest]*vertexAndLogs
+	vOrder []digest.Digest
 }
 
 type vertex struct {
-	digest digest.Digest
-	name   string
-	error  string
+	digest     digest.Digest
+	name       string
+	error      string
+	started    bool
+	completed  bool
+	cmdPrinted bool
 }
 
-type vertexLog struct {
-	vertex digest.Digest
-	msg    []byte
+// HACK: The prefix we assume here isn't valid for RUNS in exec format
+// Ex: RUN ["echo", "hello world"]
+const cmdPrefix = "/bin/sh -c "
+const buildPrefix = "    ╎ "
```

Comment:
this might be the wrong abstraction level for buildPrefix. Take a look at how we did this in Outputter. It was kind of hard to get right. And will only get more complicated if you want to manage color correctly. would be worthwhile to hide it behind an abstract io.Writer

Reply:
attn @yuindustries